velox/expression/PrestoCastHooks.cpp

```diff
@@ -65,6 +70,112 @@ Expected<int32_t> PrestoCastHooks::castStringToDate(
   return util::castFromDateString(dateString, util::ParseMode::kStandardCast);
 }
 
+namespace {
+
+// Returns an user error if 'v' is the equivalent to 'valid' but has different
+// letter cases. Returns OK if 'v' is exactly the same as 'valid' or 'v' is a
+// completely different string.
+template <TypeKind ToKind>
+Status checkInvalidCase(const StringView& v, const char* valid, size_t length) {
+  if (v.size() != length) {
+    return Status::OK();
+  }
+  bool invalidCase = false;
+  for (auto i = 0; i < length; ++i) {
+    if (v.data()[i] == valid[i]) {
+      continue;
+    }
+    if (std::tolower(v.data()[i]) == std::tolower(valid[i])) {
+      invalidCase = true;
+    } else {
+      // Not a case-insensitive stirng of 'valid'.
+      return Status::OK();
+    }
+  }
+
+  if (invalidCase) {
+    return Status::UserError(
+        "Cannot cast {} to {}", v, TypeTraits<ToKind>::name);
+  }
+  return Status::OK();
+}
+
+// Returns a user error if 'v' is equivalent to 'invalid' regardless of the
+// letter case.
+template <TypeKind ToKind>
+Status
+RejectInvalidString(const StringView& v, const char* invalid, size_t length) {
+  if (v.size() != length) {
+    return Status::OK();
+  }
+  for (auto i = 0; i < length; ++i) {
+    if (std::tolower(v.data()[i]) != std::tolower(invalid[i])) {
+      // Not a case-insensitive stirng of 'invalid'.
+      return Status::OK();
+    }
+  }
+  return Status::UserError("Cannot cast {} to {}", v, TypeTraits<ToKind>::name);
+}
+
+// Allows "NaN" but not equivalent strings with different letter cases.
+template <TypeKind ToKind>
+Status checkInvalidNaN(const StringView& v) {
+  Status result = checkInvalidCase<ToKind>(v, "NaN", 3);
+  return result;
+}
+
+// Allows "Infinity" and "-Infinity" but not equivalent strings with different
+// letter cases. Also not allows "Inf", "-Inf", and their equivalent strings.
+template <TypeKind ToKind>
+Status checkInvalidInfinity(const StringView& v) {
+  Status result = checkInvalidCase<ToKind>(v, "Infinity", 8);
+  if (!result.ok()) {
+    return result;
+  }
+  result = RejectInvalidString<ToKind>(v, "Inf", 3);
+  if (!result.ok()) {
+    return result;
+  }
+  result = checkInvalidCase<ToKind>(v, "-Infinity", 9);
+  if (!result.ok()) {
+    return result;
+  }
+  result = RejectInvalidString<ToKind>(v, "-Inf", 4);
+  return result;
+}
+
+template <typename T, TypeKind KIND>
+Expected<T> doCastToFloatingPoint(const StringView& data) {
+  auto result = util::Converter<KIND>::tryCast(data);
+  if LIKELY (!result.hasError()) {
+    if UNLIKELY (std::isnan(result.value())) {
+      auto check = checkInvalidNaN<KIND>(data);
```

Comment:
Wondering if this can be a simple string comparison:

```
if (data != "NaN") {
   return ...error;
}
```

Reply:
Ah... yes! I initially wrote checkInvalidNaN() to be called before calling util::Converter<KIND>::tryCast(), so it checks whether it's a "NaN" string with any letter case or not. Since it is called after util::Converter<KIND>::tryCast() now, it can simplified.